### ISSUE TITLE
dimulski - A malicious user can lock the prize for winning a raffle provided by the admin when a raffle is won, and deny the winner his reward

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-08-winnables-raffles-judging#150]
dimulski

High

# A malicious user can lock the prize for winning a raffle provided by the admin when a raffle is won, and deny the winner his reward

### Summary

An admin of the Winnables protocol can create a raffle by first calling the [lockNFT()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L148-L165), [lockETH()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L172-L188), or the [lockTokens()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L196-L213) functions in order to lock the prize that the winner of the raffle will receive. Then a CCIP message will be sent to the ``WinnablesTicketManager.sol`` contract, and set the **RaffleStatus** for the raffle id to **PRIZE_LOCKED**. Then the admin has to call the [createRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L252C5-L274) function in order to fully set the raffle, and users can start buying tickets. If the raffle is successful and a winner is drawn successfully, the status of the raffle will be set to 
```solidity
raffle.status = RaffleStatus.FULFILLED
```
In the [fulfillRandomWords()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L350-L361) function. Then the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function can be called by anyone with an arbitrary **prizeManager** and **chainSelector** parameters:
```solidity
    function propagateRaffleWinner(address prizeManager, uint64 chainSelector, uint256 raffleId) external {
        Raffle storage raffle = _raffles[raffleId];
        if (raffle.status != RaffleStatus.FULFILLED) revert InvalidRaffleStatus();
        raffle.status = RaffleStatus.PROPAGATED;
        address winner = _getWinnerByRequestId(raffle.chainlinkRequestId);

        _sendCCIPMessage(prizeManager, chainSelector, abi.encodePacked(uint8(CCIPMessageType.WINNER_DRAWN), raffleId, winner));
        IWinnablesTicket(TICKETS_CONTRACT).refreshMetadata(raffleId);
        unchecked {
            _lockedETH -= raffle.totalRaised;
        }
    }
```
As can be seen from the above code snippet the **raffle.status** will be set to **PROPAGATED**, and the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function will always revert when called for the same raffle ID. Raffle with **PROPAGATED** status can't be canceled as well. A malicious user can first deploy a contract that accepts CCIP messages (so the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function doesn't revert) and then call the  [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function with that contract address. Then when the winner tries to call the [claimPrize()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L105-L124) function in the ``WinnablePrizeManager.sol`` contract in order to claim his reward, the [claimPrize()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L105-L124) function will always revert as there won't be any winner set. 
```solidity
    function claimPrize(uint256 raffleId) external {
        ...
        if (msg.sender != rafflePrize.winner) revert UnauthorizedToClaim();
        ...
    }
```
The reward for the winner of the raffle will be locked forever in the ``WinnablePrizeManager.sol`` contract.

### Root Cause

The [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function can be called by anyone, with arbitrary an **prizeManager** and **chainSelector** arguments.

### Internal pre-conditions

1. Winner for raffle with ID 1 has been successfully drawn and now the status of the raffle is **PROPAGATED**, someone should call the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function in order to propagate the winner to the ``WinnablesPrizeManager.sol`` contract, and the winner can claim his prize.

### Attack Path

1. Winner for raffle with ID 1 has been successfully drawn and now the status of the raffle is **PROPAGATED**, someone should call the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function in order to propagate the winner to the ``WinnablesPrizeManager.sol`` contract, and the winner can claim his prize.
2. A malicious user deploys a contract which can receive CCIP messages, lets call it **ContractB**
3. The malicious user calls the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function but instead of providing the correct address of the ``WinnablesPrizeManager.sol`` contract, where the admin initially locked funds, he provides the address of **ContractB**. 

### Impact

The winner of the raffle won't be able to claim his prize, as there won't be any winner set for the specific raffle ID, and the [claimPrize()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L105-L124) function will always revert. The reward will be forever locked in the ``WinnablesPrizeManager.sol`` contract, the admin won't be able to withdraw it and send it manually, as the raffle can't be canceled as well. The funds that the admin locked as a reward for the winner of the raffle, either an NFT, ETH, or some ERC20 token will be locked forever

### Mitigation

Consider setting up the **prizeManager** and **chainSelector** in the constructor, or create a function that can be called only once  by the admin with the correct parameters.  Use those parameters in the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function, don't allow users to provide arbitrary parameters. 

Duplicate of #50
